File: split/badness.pas
function badness(t, s: scaled): halfword;
var
  r: integer;
begin
  if t = 0 then
    badness := 0
  else if s <= 0 then
    badness := 10000
  else
  begin
    if t <= 7230584 then
      r := (t * 297) div s
    else if s >= 1663497 then
      r := t div (s div 297)
    else
      r := t;
    if r > 1290 then
      badness := 10000
    else
      badness := (r * r * r + 131072) div 262144;
  end;
end;

File: cpp/src/badness.cpp
#include "badness.h"

//! Compute badness, given <em> t >= 0 </em>
halfword badness(scaled t, scaled s)
{
	if (t == 0)
		return 0;
	if (s <= 0)
		return inf_bad;
	// approximation to \f$\alpha t/s\f$, where \f$\alpha^3\approx100\cdot2^{18}\f$
	int r = (t <= 7230584) ? (t*297)/s : (s >= 1663497) ? t/(s/297) : t; //$297^3=99.94\times2^{18}$
	if (r > 1290)
		return inf_bad;
	return (r*r*r+1<<17)>>18; //that was \f$r^3/2^{18}\f$, rounded to the nearest integer
}
